### DISCUSSION TITLE
autoDispose -- distinction between destroying the Provider and the provider-state

### DISCUSSION THREAD

**Discussion by dgaedcke** [rrousselGit/riverpod#858] [Q&A]
I noticed while reading about autoDispose and ref.maintainState, I was confused for a bit as those seemed to be mutually exclusive ideas.

And then I realized (I hope correctly) that you're making a distinction between the provider object itself, and the state stored by it in the ProviderContainer.

And if we dispose the provider, but keep the state, what are the implications for memory leaks??   In other words, what is really gained by clearing up the provider, but leaving the state in place??

Finally, please clarify whether a NEW instance of this same provider (assuming identical Scope / ProviderContainer) will capture/reuse the stored state, or will it THEN destroy and recreate the "maintainEDState"??

**Comment by rrousselGit** [comment]
> I noticed while reading about autoDispose and ref.maintainState, I was confused for a bit as those seemed to be mutually exclusive ideas.

They aren't mutually exclusive.

`maintainState` can be changed at any time. It's not a one time thing

> And if we dispose the provider, but keep the state, what are the implications for memory leaks?? In other words, what is really gained by clearing up the provider, but leaving the state in place??

That's the opposite. Providers typically aren't disposed. It's their state that is disposed. 
There's nothing to dispose with Providers themselves. They are hardly different than a String/int. 

> Finally, please clarify whether a NEW instance of this same provider (assuming identical Scope / ProviderContainer) will capture/reuse the stored state, or will it THEN destroy and recreate the "maintainEDState"??



I don't understand this question. What does "a new instance of the same provider" mean?
